transports/quic/src/endpoint.rs

```diff
@@ -0,0 +1,813 @@
+    ) -> Result<(), quinn_proto::FinishError> {
+        self.connection.finish(id)
+    }
+
+    /// Get application-facing events
+    pub(crate) fn poll(&mut self) -> Option<quinn_proto::Event> {
+        self.connection.poll()
+    }
+
+    /// Get the side of the stream
+    pub(crate) fn side(&self) -> quinn_proto::Side {
+        self.connection.side()
+    }
+
+    /// Read from a stream
+    pub(crate) fn read(
+        &mut self,
+        id: quinn_proto::StreamId,
+        buf: &mut [u8],
+    ) -> Result<usize, quinn_proto::ReadError> {
+        self.connection.read(id, buf).map(|size| size.unwrap_or(0))
+    }
+
+    /// Close the stream, if it is not closed already
+    pub(crate) fn close(&mut self, status: u32) {
+        if self.connection.is_closed() {
+            return;
+        }
+        self.connection.close(
+            std::time::Instant::now(),
+            quinn_proto::VarInt::from_u32(status),
+            Default::default(),
+        )
+    }
+}
+
+mod channel_ref {
+    use super::{Arc, EndpointData, EndpointMessage::Dummy, Sender};
+    #[derive(Debug, Clone)]
+    pub(super) struct Channel(Sender);
+
+    impl Channel {
+        pub(super) fn new(s: Sender) -> Self {
+            Self(s)
+        }
+    }
+
+    impl std::ops::Deref for Channel {
+        type Target = Sender;
+        fn deref(&self) -> &Sender {
+            &self.0
+        }
+    }
+
+    impl std::ops::DerefMut for Channel {
+        fn deref_mut(&mut self) -> &mut Sender {
+            &mut self.0
+        }
+    }
+
+    impl Drop for Channel {
+        fn drop(&mut self) {
+            let _ = self.0.start_send(Dummy);
+        }
+    }
+
+    #[derive(Debug, Clone)]
+    pub(crate) struct EndpointRef {
+        pub(super) reference: Arc<EndpointData>,
+        // MUST COME LATER so that the endpoint driver gets notified *after* its reference count is
+        // dropped.
+        pub(super) channel: Channel,
+    }
+}
+
+pub(crate) use channel_ref::EndpointRef;
+
+/// A QUIC endpoint.  Each endpoint has its own configuration and listening socket.
+///
+/// You generally need only one of these per process.  Endpoints are [`Send`] and [`Sync`], so you
+/// can share them among as many threads as you like.  However, performance may be better if you
+/// have one per CPU core, as this reduces lock contention.  Most applications will not need to
+/// worry about this.  `Endpoint` tries to use fine-grained locking to reduce the overhead.
+///
+/// `Endpoint` wraps the underlying data structure in an [`Arc`], so cloning it just bumps the
+/// reference count.  All state is shared between the clones.  For example, you can pass different
+/// clones to [`Transport::listen_on`].  Each incoming connection will be received by exactly one of
+/// them.
+///
+/// The **only** valid [`Multiaddr`] to pass to `listen_on` is the one used to create the
+/// `QuicEndpoint`.  You can obtain this via the `addr` method.  If you pass a different one, you
```

Comment:
```suggestion
/// `Endpoint`.  You can obtain this via the `addr` method.  If you pass a different one, you
```

I am guessing you meant `Endpoint`? Might be worth using doc links as they can be validated at _compile_ time.

Reply:
Done.